ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,187 @@
+
+  private final int pruningLimit;
+  private final int loadingLimit;
+  private final BonsaiWorldStateKeyValueStorage rootWorldStateStorage;
+  private final Blockchain blockchain;
+  private final long numBlocksToRetain;
+  private static final Multimap<Long, byte[]> knownTrieLogKeysByDescendingBlockNumber =
+      TreeMultimap.create(Comparator.reverseOrder(), Comparator.comparingInt(Arrays::hashCode));
+
+  public TrieLogPruner(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final Blockchain blockchain,
+      final long numBlocksToRetain,
+      final int pruningLimit) {
+    this.rootWorldStateStorage = rootWorldStateStorage;
+    this.blockchain = blockchain;
+    this.numBlocksToRetain = numBlocksToRetain;
+    this.pruningLimit = pruningLimit;
+    this.loadingLimit = pruningLimit; // same as pruningLimit for now
+  }
+
+  public void initialize() {
+    preloadCache();
+  }
+
+  private void preloadCache() {
+    LOG.atInfo()
+        .setMessage("Loading first {} trie logs from database...")
+        .addArgument(loadingLimit)
+        .log();
+    final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit);
+    final AtomicLong count = new AtomicLong();
+    trieLogKeys.forEach(
+        blockHashAsBytes -> {
+          Hash hash = Hash.wrap(Bytes32.wrap(blockHashAsBytes));
+          final Optional<BlockHeader> header = blockchain.getBlockHeader(hash);
+          if (header.isPresent()) {
+            knownTrieLogKeysByDescendingBlockNumber.put(header.get().getNumber(), blockHashAsBytes);
+            count.getAndIncrement();
+          } else {
+            // prune orphaned blocks (sometimes created during block production)
+            rootWorldStateStorage.pruneTrieLog(blockHashAsBytes);
+          }
+        });
+    LOG.atInfo().log("Loaded {} trie logs from database", count);
+    pruneFromCache();
+  }
+
+  void cacheForLaterPruning(final long blockNumber, final byte[] trieLogKey) {
+    LOG.atTrace()
+        .setMessage("caching trie log for later pruning blockNumber {}; trieLogKey (blockHash) {}")
+        .addArgument(blockNumber)
+        .addArgument(Bytes.wrap(trieLogKey).toHexString())
+        .log();
+    knownTrieLogKeysByDescendingBlockNumber.put(blockNumber, trieLogKey);
+  }
+
+  void pruneFromCache() {
+
+    final long retainAboveThisBlock = blockchain.getChainHeadBlockNumber() - numBlocksToRetain;
+    final long retainAboveThisBlockOrFinalized =
+        blockchain
+            .getFinalized()
+            .flatMap(blockchain::getBlockHeader)
+            .map(ProcessableBlockHeader::getNumber)
+            .map(finalizedBlock -> Math.min(finalizedBlock, retainAboveThisBlock))
+            .orElse(retainAboveThisBlock);
+
+    LOG.atTrace()
+        .setMessage(
+            "min((chainHeadNumber: {} - numBlocksToRetain: {}), finalized: {})) = retainAboveThisBlockOrFinalized: {}")
+        .addArgument(blockchain.getChainHeadBlockNumber())
```

Comment:
maybe we can call this method only one time https://github.com/hyperledger/besu/pull/6026/files#diff-60f4623b862cc8abfc075f66848a976b23470f2d6f1ce1124773835116594110R105

Reply:
true, I figured it would only get called during trace logging, but still.

Reply:
chainHead is cached so it's not a db call, but I've made sure we're only executing it when trace is enabled...
https://github.com/hyperledger/besu/pull/6026/commits/c74b104fc0779f8e8375b0f3e63e7e3e85d710b4